File: Recursion/fibonacci_memoization.cpp
/*
	Implelement memoized fibonacci

	Memoization is an enhancement procedure used to speed up computer programs by keeping the values
	of distinct function calls and returning the stored input when the same function is invoked again.

	Sample Input: 6
	Output: 8
*/
#include<bits/stdc++.h>
using namespace std;
int F[51];
int fib(int n){
	if(F[n] != -1)
		return F[n];
	F[n] = fib(n - 1) + fib(n-2);	
	return F[n];
}
int main(){	
	for(int i = 0; i < 51; i++)
		F[i] = -1;
	F[0] = 0;
	F[1] = 1;	
	int n;
	cout << "Give me n : ";
	cin >> n;
	int result = fib(n);
	cout << result;
return 0;
}

File: Recursion/fibonacci_memoization.go
/*
	Implelement memoized fibonacci

	Memoization is an enhancement procedure used to speed up computer programs by keeping the values
	of distinct function calls and returning the stored input when the same function is invoked again.

	Sample Input: 6
	Output: 8
*/
package main

import "fmt"

var F = make([]int, 51)
// array based implementation 0th fib = 1
func Fibonacci(n int) int {	
	if F[n] != -1 {
		return F[n]
	}
	F[n] = Fibonacci(n - 1) + Fibonacci(n - 2)
	return F[n]
}
// more optimized 1st fibonacci = 0, second = 1
func GetNthFib(n int) int {
    firstNo := 0
    secondNo := 1
    result := 0
    if n <= 2 {
        return n - 1
    } 
    for i := 2; i < n; i++ {
		// move first and second to to next iteration
        result = firstNo + secondNo
        firstNo = secondNo
        secondNo = result
    }
	return result
}
func main() {
	for i := 0; i < 51; i++ {
		F[i] = -1
	}
	F[0] = 0
	F[1] = 1
	msg := Fibonacci(5)
	fmt.Println(msg)
	msg = GetNthFib(5)
	fmt.Println(msg)
}